Mathlib/Algebra/Polynomial/SumIteratedDerivative.lean

```diff
@@ -0,0 +1,253 @@
+        C ((i + k).descFactorial k • p.coeff (i + k)) * X ^ i := by rw [coeff_iterate_derivative]
+    _ = C ((k ! * (i + k).choose k) • p.coeff (i + k)) * X ^ i := by
+      rw [Nat.descFactorial_eq_factorial_mul_choose]
+    _ = (k ! * (i + k).choose k) • C (p.coeff (i + k)) * X ^ i := by rw [smul_C]
+    _ = k ! • (i + k).choose k • C (p.coeff (i + k)) * X ^ i := by rw [mul_smul]
+    _ = k ! • ((i + k).choose k • C (p.coeff (i + k)) * X ^ i) := by rw [smul_mul_assoc]
+
+end Semiring
+
+section CommRing
+
+variable [CommRing R] {A : Type*} [CommRing A] [Algebra R A]
+
+theorem iterate_derivative_small (p : R[X]) (q : ℕ) (r : A) {p' : A[X]}
+    (hp : p.map (algebraMap R A) = (X - C r) ^ q * p') {k : ℕ} (hk : k < q) :
+    aeval r (derivative^[k] p) = 0 := by
+  have h (x) : (X - C r) ^ (q - (k - x)) = (X - C r) ^ 1 * (X - C r) ^ (q - (k - x) - 1) := by
+    rw [← pow_add, add_tsub_cancel_of_le]
+    rw [Nat.lt_iff_add_one_le] at hk
+    exact (le_tsub_of_add_le_left hk).trans (tsub_le_tsub_left (tsub_le_self : _ ≤ k) _)
+  rw [aeval_def, eval₂_eq_eval_map, ← iterate_derivative_map]
+  simp_rw [hp, iterate_derivative_mul, iterate_derivative_X_sub_pow, ← smul_mul_assoc, smul_smul,
+    h, ← mul_smul_comm, mul_assoc, ← mul_sum, eval_mul, pow_one, eval_sub, eval_X, eval_C, sub_self,
+    zero_mul]
+
+theorem iterate_derivative_of_eq (p : R[X]) (q : ℕ) (r : A) {p' : A[X]}
+    (hp : p.map (algebraMap R A) = (X - C r) ^ q * p') :
+    aeval r (derivative^[q] p) = q ! • p'.eval r := by
+  have h (x) (h : 1 ≤ x) (h' : x ≤ q) :
+      (X - C r) ^ (q - (q - x)) = (X - C r) ^ 1 * (X - C r) ^ (q - (q - x) - 1) := by
+    rw [← pow_add, add_tsub_cancel_of_le]
+    rwa [tsub_tsub_cancel_of_le h']
+  rw [aeval_def, eval₂_eq_eval_map, ← iterate_derivative_map]
+  simp_rw [hp, iterate_derivative_mul, iterate_derivative_X_sub_pow, ← smul_mul_assoc, smul_smul]
+  rw [sum_range_succ', Nat.choose_zero_right, one_mul, tsub_zero, Nat.descFactorial_self, tsub_self,
+    pow_zero, smul_mul_assoc, one_mul, Function.iterate_zero_apply, eval_add, eval_smul]
+  convert zero_add _
+  rw [eval_finset_sum]
+  apply sum_eq_zero
+  intro x hx
+  rw [h (x + 1) le_add_self (Nat.add_one_le_iff.mpr (mem_range.mp hx)), pow_one,
+    eval_mul, eval_smul, eval_mul, eval_sub, eval_X, eval_C, sub_self, zero_mul,
+    smul_zero, zero_mul]
+
+variable (A)
+
+theorem iterate_derivative_large (p : R[X]) (q : ℕ) {k : ℕ} (hk : q ≤ k) :
+    ∃ gp : R[X], gp.natDegree ≤ p.natDegree - k ∧
+      ∀ r : A, aeval r (derivative^[k] p) = q ! • aeval r gp := by
+  obtain ⟨p', p'_le, hp'⟩ := iterate_derivative_eq_factorial_mul p k
+  obtain ⟨k, rfl⟩ := Nat.exists_eq_add_of_le hk
+  refine ⟨((q + k).descFactorial k : R[X]) * p', (natDegree_C_mul_le _ _).trans p'_le, fun r => ?_⟩
+  simp_rw [hp', nsmul_eq_mul, map_mul, map_natCast, ← mul_assoc, ← Nat.cast_mul,
+    Nat.add_descFactorial_eq_ascFactorial, Nat.factorial_mul_ascFactorial]
+
+theorem sumIderiv_sl (p : R[X]) (q : ℕ) :
```

Comment:
I don't understand the `sl` suffix of this name.

Reply:
@FR-vdash-bot do you recall?